README.md
# :warning: THIS REPO HAS BEEN ARCHIVED :warning:

This repository contains a custom plugin for WordPress sites formerly under the america.gov domain. Theses sites were transitioned to another team for maintenance in October 2020. As such, this repository has been archived and is no longer being maintained. The code is preserved here as a reference.

# IIP Maps

This plugin adds a custom map post type to a WordPress site.

The admin pulls in events from Screendoor and stores them in the database.

## Developers

For developer documentation please see the [docs](./docs/index.md).


CHANGELOG.md
# Change Log

**All notable changes to this project will be documented in this file.**

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased](https://github.com/IIP-Design/iip-map/compare/v2.1.2...HEAD)

_This sections lists changes committed since most recent release_

## [v2.1.2](https://github.com/IIP-Design/iip-map/compare/v2.1.1...v2.1.2) - 2020-08-23

### Fixed

- Downgrade ol package back down to version 5 since upgrade to 6 broke map rendering in admin page

## [v2.1.1](https://github.com/IIP-Design/iip-map/compare/v2.1.0...v2.1.1) - 2020-08-21

### Fixed

- Convert supports values in `register_post_type` function to an array to avoid errors in WordPress versions > 5.3

### Security

- Updated dependencies

## [v2.1.0](https://github.com/IIP-Design/iip-map/compare/v2.0.0...v2.1.0) - 2019-09-19

### Added

- A filter to remove exported events from the table list

## [v2.0.0](https://github.com/IIP-Design/iip-map/compare/v1.1.4...v2.0.0) - 2019-07-03

### Added

- Filtering of markers based on topic and date
- Visual representation of map in the admin panel to help with generating shortcodes
- Customizable map cards
- Table view of events below the map
- Marker editing based off of Screendoor id

### Changed

- Add plugin deactivation hook to give option to delete all project and map data
- Presentation layer and geocoding switched to MapBox

## [v1.1.4](https://github.com/IIP-Design/iip-map/compare/v1.1.3...v1.1.4) - 2018-06-26

### Changed

- Revert back to full OpenLayers library

### Fixed

- Use ISO date format to avoid "Invalid Date" and `NaN` errors in Safari and mobile browsers

## [v1.1.3](https://github.com/IIP-Design/iip-map/compare/v1.1.2...v1.1.3) - 2018-06-21

### Changed

- Pull out only required methods from the OpenLayers library to reduce the amount of JavaScript loaded onto the page

### Fixed

- Add 0 to the front of single digit minutes in event time value

## [v1.1.2](https://github.com/IIP-Design/iip-map/compare/v1.1.1...v1.1.2) - 2018-06-19

### Changed

- Add visual cues to indicate that OpenLayers InfoWindow is an accordion when more than one event listed

## [v1.1.1](https://github.com/IIP-Design/iip-map/compare/v1.1.0...v1.1.1) - 2018-06-13

### Changed

- Recenter map on InfoWindow when clicking on marker (in OpenLayers and Google Maps)
- Reorder information on InfoWindows and make contact block conditional

### Fixed

- Venue address in geocoder pulling wrong post metadata value

## [v1.1.0](https://github.com/IIP-Design/iip-map/compare/v1.0.3...v1.1.0) - 2018-06-06

### Added

- Map type dropdown to shortcode generator
- Ability to edit the location of or entirely delete existing markers
- InfoWindow popups and clustering on OpenLayers maps

### Changed

- Set event duration field to accept radio options rather than text input

### Fixed

- When set to noon, event time was being converted to midnight

## [v1.0.3](https://github.com/IIP-Design/iip-map/compare/v1.0.2...v1.0.3) - 2018-06-01

### Added

- Distinct API endpoints allowing for multiple map projects on one site
- Option to display map using OpenLayers rather than Google Maps

### Changed

- Display date in InfoWindow in 'Month DD' format rather than YYYY-MM-DD

### Fixed

- Force download of file with data when exporting project data
- Venue address was displaying the wrong value in the administrative panel

## [v1.0.2](https://github.com/IIP-Design/iip-map/compare/v1.0.1...v1.0.2) - 2018-05-25

### Changed

- Increase geocoding batch size from 10 responses to 100 responses
- Switch from fetch to an AJAX call when retrieving map data (promise was throwing CORS error)

## [v1.0.1](https://github.com/IIP-Design/iip-map/compare/v1.0.0...v1.0.1) - 2018-05-24

### Added

- Host name an event parameter

### Fixed

- Allow for non-default table prefixes when generating API endpoint
- Add CSS rule to prevent zoom and other Google map icons from appearing blurry
- Add closing `<div>` tag to shortcode output to allow for content to be added below the map

### Security

- Add nonces to internal AJAX calls

## [v1.0.0](https://github.com/IIP-Design/iip-map/tree/v1.0.0) - 2018-05-24 (Initial Release)

### Added

- Map data custom post type with metadata saved to custom table
- Admin dashboard to configure map projects and generate map shortcodes
- Ability to import project data via the [Screendoor API](http://dobtco.github.io/screendoor-api-docs/)
- Custom WordPress API endpoint to relay map data to embedded map
- Custom WordPress shortcode to embed map projects on posts
- Frontend rendering of map, with marker clustering and spidering


version.js
const chalk = require("chalk");
const git = require("simple-git");
const gitRev = require("git-rev-sync");
const replace = require("replace");

module.exports.versionBump = () => {
  // Get the type of version bump (major, minor, patch) from command argument
  const type = process.argv[1];

  // Get current git tag and status
  const version = gitRev.tag();
  const isDirty = gitRev.isDirty();

  // Check for uncommitted changes and throw error if present
  if (isDirty) {
    console.log(
      chalk.red("ERROR:"),
      "You have uncommitted changes on your branch.",
      "\nPlease commit all changes before attempting to version your plugin.\n"
    );
    return;
  }

  // Split the git version into it's component parts
  const split = version.split(".");
  let prefix = split[0].includes("v") ? "v" : "";
  let major = split[0].includes("v") ? split[0].slice(1) : split[0];
  let minor = split[1];
  let patch = split[2];

  // Update version based on the version type passed in the command
  if (type === "major") {
    major = Number(major) + 1;
    minor = 0;
    patch = 0;
  } else if (type === "minor") {
    minor = Number(minor) + 1;
    patch = 0;
  } else if (type === "patch") {
    patch = Number(patch) + 1;
  } else {
    console.log(
      chalk.red("ERROR:"),
      "Please provide a valid version type as an argument.",
      "\n(For example: `npm run version patch`)",
      '\nAccepted values are "major", "minor", and "patch"',
      "\nFor more information see https://semver.org/\n"
    );
    return;
  }
  const newVersion = `${prefix}${major}.${minor}.${patch}`;

  console.log("Updating the plugin version number in the following locations:");

  // Sets the new version in the plugin header
  replace({
    regex: `Version: ${version}`,
    replacement: `Version: ${newVersion}`,
    paths: ["./iip-map.php"]
  });

  // Sets the new version in hook registration
  replace({
    regex: `version = '${version}'`,
    replacement: `version = '${newVersion}'`,
    paths: ["./includes/class-iip-map.php"]
  });

  // Sets the new version in the main, admin, and frontend package.json files
  replace({
    regex: `"version": "${version}"`,
    replacement: `"version": "${newVersion}"`,
    paths: [
      "./package.json",
      "./admin/js/package.json",
      "./public/package.json"
    ]
  });

  // Commit version bump and tag the branch
  git()
    .add("./*")
    .commit(`${newVersion} Release`)
    .addTag(newVersion);
};


iip-map.php
<?php
/**
 * Plugin Name: Map Embed for IIP Properties
 * Description: Adds ability to embed a Google map onto a WordPress site
 * Version: v2.1.2
 * Author: Marek Rewers
 * Text Domain: iip-map
 * License: GPLv2 or laters
 */

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
 	die;
}

// Define constants
define( 'IIP_MAP_DIR', plugin_dir_path( dirname( __FILE__ ) ) . 'iip-map/' );
define( 'IIP_MAP_URL', plugin_dir_url( dirname( __FILE__ ) ) . 'iip-map/' );

global $wpdb;
define( 'IIP_MAP_EVENTS_TABLE', "{$wpdb->prefix}iip_map_data" );
define( 'IIP_MAP_GEOCODER_BATCH_SIZE', 20 );

function iip_map_activate() {
	require_once plugin_dir_path( __FILE__ ) . 'includes/class-iip-map-activator.php';
	IIP_Map_Activator::activate();
}
register_activation_hook( __FILE__, 'iip_map_activate' );
register_activation_hook( __FILE__, 'iip_map_create_db' );


function iip_map_deactivate() {
	require_once plugin_dir_path( __FILE__ ) . 'includes/class-iip-map-deactivator.php';
	IIP_Map_Deactivator::deactivate();
}
register_deactivation_hook( __FILE__, 'iip_map_deactivate' );

require plugin_dir_path( __FILE__ ) . 'includes/class-iip-map.php';

/* Begin execution of the plugin.
 *
 * Since everything within the plugin is registered via hooks,
 * then kicking off the plugin from this point in the file does
 * not affect the page life cycle.
 *
 */

function run_iip_map() {
 	$plugin = new IIP_Map();
 	$plugin->run();
}

run_iip_map();


public/embed-map.php
<?php

// Embedded Google Map on page
class IIP_Map_Embed {

  public function __construct( $plugin_name, $version ) {
    $this->plugin_name = $plugin_name;
    $this->version = $version;
  }

  // Register script that embeds the map
  public function iip_map_register_embed() {
    wp_register_script( 'draw-gmap', plugin_dir_url( __FILE__ ) . 'js/dist/draw-gmap.min.js', array(), null, true );

    wp_register_script( 'draw-ol-map', plugin_dir_url( __FILE__ ) . 'js/dist/draw-ol-map.min.js', array(), null, true );

    wp_register_script( 'openlayers', plugin_dir_url( __FILE__ ) . 'js/dist/ol.js', array(), null, true );

    wp_register_script( 'marker-clusterer', plugin_dir_url( __FILE__ ) . 'js/dist/markerclusterer.min.js', array(), null, false);

    wp_register_script( 'marker-spiderfy', plugin_dir_url( __FILE__ ) . 'js/dist/overlapping-marker-spiderfy.min.js', array(), null, true);

    wp_enqueue_style( 'iip-map-frontend', plugin_dir_url( __FILE__ ) . 'css/iip-map-frontend.css', array(), $this->version, 'all' );

    wp_enqueue_style( 'ol-frontend', plugin_dir_url( __FILE__ ) . 'css/ol.css', array(), $this->version, 'all' );

	wp_register_script( 'table-button', plugin_dir_url( __FILE__ ) . 'js/dist/table-button.min.js', array(), null, true );

	wp_register_script( 'event-table', plugin_dir_url( __FILE__ ) . 'table/js/dist/table-app.js', array(), null, true );

	wp_enqueue_style( 'iip-table-frontend', plugin_dir_url( __FILE__ ) . 'table/css/table.css', array(), $this->version, 'all' );
  }

  // The output of the map shortcode
  public function iip_map_shortcode( $args ) {
    $attr = shortcode_atts( array(
      'id'     => '',
      'height' => 600,
      'zoom'   => 2,
      'lat'    => 30,
      'lng'    => 0,
      'type'   => 'ol'
    ), $args );

    // Set shortcode attributes as variables
    $map = $attr['id'];
    $height = $attr['height'];
    $zoom = $attr['zoom'];
    $lat = $attr['lat'];
    $lng = $attr['lng'];
    $type = $attr['type'];

    // Pass variables to map drawing file (for OpenLayers)
    wp_localize_script( 'draw-ol-map', 'iip_map_params', array(
      'map_id' => $map,
      'map_zoom' => $zoom,
      'map_center_lat' => $lat,
      'map_center_lng' => $lng
    ));

    // Pass variables to map drawing file (for Google Maps)
    wp_localize_script( 'draw-gmap', 'iip_map_params', array(
      'google_api_key' => get_option( 'iip_map_google_maps_api_key' ),
      'map_id' => $map,
      'map_zoom' => $zoom,
      'map_center_lat' => $lat,
      'map_center_lng' => $lng
    ));

    if ($type == 'ol' || $type == '') {
      // Load MarkerClusterer and return map
      wp_enqueue_script( 'openlayers' );
      wp_enqueue_script( 'draw-ol-map' );
	  wp_enqueue_script( 'table-button' );
	  wp_enqueue_script( 'event-table' );
    } elseif ($type == 'gmap') {
      // Load MarkerClusterer and return map
      wp_enqueue_script( 'marker-clusterer' );
      wp_enqueue_script( 'draw-gmap' );
      wp_enqueue_script( 'marker-spiderfy' );
      wp_enqueue_script( 'table-button' );
      wp_enqueue_script( 'event-table' );
    }

    $html = '<div id="map" style="height: ' . $height . 'px" class="iip-map-container" data-map-id="' . $map . '"><div id="popup"></div></div><button onclick="toggleTable()" id="toggle-table" data-text-original="View this map as a table" data-text-swap="Hide table">View this map as a table</button><div id="event-list" style="display: none;"></div>';
    return $html;

  }

  // Register the map shortcode
  public function iip_map_add_shortcode() {
    add_shortcode( 'map', array( $this, 'iip_map_shortcode' ) );
  }
}


includes/class-iip-map.php
<?php

class IIP_Map {

  /**
   * The loader that's responsible for maintaining and registering all hooks that power the plugin.
   *
   * @since    1.0.0
   * @access   protected
   * @var      IIP_Map_Loader    $loader    Maintains and registers all hooks for the plugin.
   */

  protected $loader;

  /**
   * The unique identifier  and version of this plugin.
   *
   * @since    1.0.0
   * @access   protected
   */

  protected $plugin_name;
  protected $version;

  /**
   * Define the core functionality of the plugin.
   *
   * Set the plugin name and the plugin version that can be used throughout the plugin.
   * Load the dependencies, define the locale, and set the hooks for the admin area and
   * the public-facing side of the site.
   *
   * @since    1.0.0
   */
  public function __construct() {
    $this->plugin_name = 'iip-map';
    $this->version = 'v2.1.2';

    $this->load_dependencies();
    $this->define_admin_hooks();
    $this->define_public_hooks();
  }

  /**
   * Load the required dependencies for this plugin.
   *
   * Include the following files that make up the plugin:
   *
   * - IIP_Map_Loader. Orchestrates the hooks of the plugin.
   * - IIP_Map_Admin. Defines all hooks for the admin area.
   * - IIP_Map_Embed. Defines all hooks for the public side of the site.
   *
   * Create an instance of the loader which will be used to register the hooks with WordPress.
   *
   * @since    1.0.0
   * @access   private
   */

  private function load_dependencies() {

    // The class responsible for orchestrating the actions and filters of the core plugin.
    require_once plugin_dir_path( dirname( __FILE__ ) ) . 'includes/class-iip-map-loader.php';

    // The class responsible for defining all actions that occur in the admin area.
    require_once plugin_dir_path( dirname( __FILE__ ) ) . 'admin/class-iip-map-admin.php';
    require_once plugin_dir_path( dirname( __FILE__ ) ) . 'admin/map-data-post-type.php';

    require_once plugin_dir_path( dirname( __FILE__ ) ) . 'admin/ajax/geocoder.php';
    require_once plugin_dir_path( dirname( __FILE__ ) ) . 'admin/ajax/map-export.php';
    require_once plugin_dir_path( dirname( __FILE__ ) ) . 'admin/ajax/map-import.php';
    require_once plugin_dir_path( dirname( __FILE__ ) ) . 'admin/ajax/marker-update.php';
    require_once plugin_dir_path( dirname( __FILE__ ) ) . 'admin/ajax/save-events.php';
    require_once plugin_dir_path( dirname( __FILE__ ) ) . 'admin/ajax/save-fields.php';

    // The class responsible for defining all actions that occur in the public-facing side of the site.
    require_once plugin_dir_path( dirname( __FILE__ ) ) . 'public/embed-map.php';
    require_once plugin_dir_path( dirname( __FILE__ ) ) . 'public/generate-api.php';

    $this->loader = new IIP_Map_Loader();
  }

  // Register all of the hooks related to the admin area functionality of the plugin.
  private function define_admin_hooks() {
    $plugin_admin = new IIP_Map_Admin( $this->get_plugin_name(), $this->get_version() );
    $plugin_export_ajax = new IIP_Map_Export();
    $plugin_import_ajax = new IIP_Map_Import();
    $plugin_update_ajax = new IIP_Map_Marker_Update();
    $plugin_fields_ajax = new IIP_Map_Save_Fields();
    $plugin_events_ajax = new IIP_Map_Save_Events();
    $plugin_geocode_ajax = new IIP_Map_Geocoder();
    $plugin_post_type = new IIP_Map_Post_Type();

    // Admin hooks
    $this->loader->add_action( 'admin_init', $plugin_admin, 'iip_map_settings_sections' );
    $this->loader->add_action( 'admin_init', $plugin_admin, 'iip_map_settings_fields' );
    $this->loader->add_action( 'admin_menu', $plugin_admin, 'iip_map_admin_menu' );
    $this->loader->add_action( 'admin_notices', $plugin_admin, 'iip_map_localize_variables' );
    $this->loader->add_action( 'admin_enqueue_scripts', $plugin_admin, 'iip_map_admin_enqueue' );
    // Ajax hooks
    $this->loader->add_action( 'wp_ajax_iip_map_import_data_ajax', $plugin_import_ajax, 'import_data_ajax' );
    $this->loader->add_action( 'wp_ajax_iip_map_export_data_ajax', $plugin_export_ajax, 'export_data_ajax' );
    $this->loader->add_action( 'wp_ajax_iip_map_save_fields_ajax', $plugin_fields_ajax, 'save_fields_ajax' );
    $this->loader->add_action( 'wp_ajax_iip_map_save_events_ajax', $plugin_events_ajax, 'save_events_ajax' );
    $this->loader->add_action( 'wp_ajax_iip_map_geocode_events_ajax', $plugin_geocode_ajax, 'geocode_events_ajax' );
    $this->loader->add_action( 'wp_ajax_iip_map_get_marker_ajax', $plugin_update_ajax, 'get_marker_ajax' );
    $this->loader->add_action( 'wp_ajax_iip_map_update_marker_ajax', $plugin_update_ajax, 'update_marker_ajax' );
    $this->loader->add_action( 'wp_ajax_iip_map_delete_marker_ajax', $plugin_update_ajax, 'delete_marker_ajax' );
    // Post type hooks
    $this->loader->add_action( 'init', $plugin_post_type, 'create_map_post_type' );
    $this->loader->add_action( 'save_post', $plugin_post_type, 'save_map_meta', 10, 2 );
  }

  // Register all of the hooks related to the public-facing functionality
  private function define_public_hooks() {
    $plugin_map = new IIP_Map_Embed( $this->get_plugin_name(), $this->get_version() );
    $plugin_api = new IIP_Map_API_Route();

    $this->loader->add_action( 'init', $plugin_map, 'iip_map_register_embed' );
    $this->loader->add_action( 'init', $plugin_map, 'iip_map_add_shortcode' );
    $this->loader->add_action( 'rest_api_init', $plugin_api, 'create_map_endpoint' );
  }

  /**
   * Run the loader to execute all of the hooks with WordPress.
   *
   * @since    1.0.0
   */

  public function run() {
    $this->loader->run();
  }

  /**
   * The reference to the class that orchestrates the hooks with the plugin.
   *
   * @since     1.0.0
   * @return    IIP_Map_Loader    Orchestrates the hooks of the plugin.
   */

  public function get_loader() {
    return $this->loader;
  }

  // Retrieve the name & version number of the plugin.
  public function get_plugin_name() {
    return $this->plugin_name;
  }

  public function get_version() {
    return $this->version;
  }
}


public/table/js/src/Components/Table.js
import React, { Component } from 'react';
// Import shortcode parameters
const mapID = iip_map_params.map_id; // eslint-disable-line no-undef, camelcase
const baseURL = '/wp-json/iip-map/v1/maps/' + mapID; // eslint-disable-line prefer-template

class Table extends Component {
  state = {
    isLoading: true,
    filter: '',
    events: [],
    error: null
  };

  componentDidMount() {
    this.fetchEvents();
  }

  handleChange = ( event ) => {
    this.setState( { filter: event.target.value } );
  };

  fetchEvents() {
    fetch( baseURL )
      .then( response => response.json() )
      .then( data => this.setState( {
        events: data,
        isLoading: false
      } ) )
      .catch( error => this.setState( { error, isLoading: false } ) );
  }

  render() {
    const {
      isLoading, events, error, filter
    } = this.state;
    const lowercasedFilter = filter.toLowerCase();
    const filteredData = events.filter( ( event ) => { // eslint-disable-line arrow-body-style
      return Object.keys( event ).some( key => event[key].toLowerCase().includes( lowercasedFilter ) );
    } );

    return (
      <React.Fragment>
        <input className="event-table__filter" placeholder="Filter results..." value={ filter } onChange={ this.handleChange } />
        <div className="event-table">
          { error ? <p>{ error.message }</p> : null }
          { !isLoading ? (
            <React.Fragment>
              <div className="column-labels">
                <div className="label">Event Name</div>
                <div className="label">Event Date</div>
                <div className="label">Venue Address</div>
                <div className="label">Contact Info</div>
              </div>
            </React.Fragment>
          ) : null }
          { !isLoading ? (
            filteredData.map( ( event ) => {
              const cdate = ( new Date( event.event_date ).toLocaleDateString() ); // eslint-disable-line camelcase
              return (
                <div className="event-table__row event">
                  <h3>{ event.event_name }</h3>
                  <div className="date">{ cdate }</div>
                  <div className="location">
                    { event.venue_address }
                    { ', ' }
                    { event.venue_region }
                  </div>
                  <div className="contact">{ event.contact }</div>
                </div>
              );
            } )
          ) : (
            <h3>Loading...</h3>
          ) }
        </div>
      </React.Fragment>
    );
  }
}

export default Table;


public/src/js/draw-map.js
import mapboxgl from 'mapbox-gl';
// ie 11 compatibility
import 'isomorphic-fetch';
import { polyfill } from 'es6-promise';

polyfill();

// shortcode parameters
const mapId = iip_map_params.map_id; // eslint-disable-line no-undef, camelcase
const apiKey = iip_map_params.mapbox_api_key; // eslint-disable-line no-undef, camelcase
const mapZoom = iip_map_params.map_zoom; // eslint-disable-line no-undef, camelcase
const lat = iip_map_params.map_center_lat; // eslint-disable-line no-undef, camelcase
const lng = iip_map_params.map_center_lng; // eslint-disable-line no-undef, camelcase
const { mapping } = iip_map_params; // eslint-disable-line no-undef, camelcase
const { card } = iip_map_params; // eslint-disable-line no-undef, camelcase

// get topic select
const topicSelect = document.getElementById( 'topic-select' );
const fragment = document.createDocumentFragment();
// get past events checkbox
const pastEventsCheckbox = document.getElementById( 'past-events-checkbox' );
// set to store events for cluster filtering
const storage = new Set();

// today's date
const todaysDate = new Date();
// we are making the assumption that events take place in the current year
const currentYear = todaysDate.getFullYear();

mapboxgl.accessToken = apiKey;

const map = new mapboxgl.Map( {
  container: 'map',
  style: 'mapbox://styles/alexgordon/cjxgmh0p22nvk1cny8rtlluxg?optimize=true&fresh=true',
  center: [lat, lng],
  zoom: mapZoom
} );

// Configure Map controls
// Add zoom and rotation controls to the map.
const mapNav = new mapboxgl.NavigationControl( { showCompass: false } );
map.addControl( mapNav, 'bottom-right' );

// Add custom pin to map
map.loadImage( '/wp-content/plugins/iip-map/public/images/location-pin.png', ( error, image ) => {
  if ( error ) {
    throw error;
  }
  map.addImage( 'pin', image );
} );

const mashObject = ( obj ) => {
  if ( Array.isArray( obj ) ) {
    return obj.join( ', ' );
  }
  if ( typeof obj === 'object' ) {
    const strs = [];
    Object.keys( obj ).forEach( ( key ) => {
      strs.push( `${key}: ${mashObject( obj[key] )}` );
    } );
    return strs.join( '\n' );
  }
  return obj.toString();
};

// map mapped fields to available fields based on screendoor ID
const parseSection = ( sectionMap, fields ) => {
  const vals = [];

  sectionMap.forEach( ( { fieldType, field } ) => {
    // http://dobtco.github.io/screendoor-api-docs/#spec-for-the-response-hash
    switch ( fieldType ) {
      case 'text':
      case 'paragraph':
      case 'dropdown':
      case 'email':
      case 'phone':
      case 'numeric':
      case 'website':
        if ( field in fields && fields[field] ) {
          vals.push( fields[field].replace( /^[ \t\r\n]+|[ \t\r\n]+$/g, '' ) );
        }
        break;
      case 'date':
      case 'time':
      case 'address':
      case 'price':
        if ( field in fields && fields[field] ) {
          vals.push( fields[field] );
        }
        break;
      case 'radio':
        if ( field in fields && fields[field] ) {
          vals.push( fields[field] );
        }
        break;
      default:
        if ( field in fields && fields[field] ) {
          vals.push( fields[field] );
        }
        return '';
    }
  } );
  return vals;
};

// build filter and associated functionality
// store layers in browser
function buildFilter( m ) {
  const layersUnique = new Set();
  let mapData = m;
  // mapData is returned as a string ie11 so we need to convert it to an object
  if ( typeof mapData === 'string' ) {
    mapData = JSON.parse( m );
  }

  mapData.features.forEach( ( marker ) => {
    const eventTopic = marker.properties.topic;
    const layerID = `${eventTopic}`;
    layersUnique.add( layerID );
  } );

  layersUnique.forEach( ( layerID ) => {
    // we are going to build arrays of markers to store filtered by layerID
    const markers = [];
    mapData.features.forEach( ( marker ) => {
      // build array of markers filter by layerID
      if ( marker.properties.topic === layerID ) {
        // push markers to array
        markers.push( marker );
      }
    } );
    // store events sorted by category
    // we swap the map data with stored data if the filter is used
    // doing this because we cannot update the cluster layer once it's drawn
    storage[layerID] = markers;
    // build out filter <select> by adding each layerID as <option>
    const option = document.createElement( 'option' );
    option.innerHTML = layerID;
    option.value = layerID;
    fragment.appendChild( option );
  } );
  // build filtering logic which is shared between checkbox and select events
  function filterLogic() {
    // get all events if select is empty
    let selectValue = topicSelect.value;
    if ( !selectValue ) {
      selectValue = 'all';
    }
    map.getSource( 'events' ).setData( {
      type: 'FeatureCollection',
      features: storage[selectValue].filter( ( marker ) => {
        if ( pastEventsCheckbox.checked ) {
          const { fields } = marker.properties;
          const dateField = parseSection( mapping.date_arr, fields );
          if ( dateField.length < 1 || !dateField[0].month || !dateField[0].day ) {
            return true;
          }
          const eventDate = new Date( currentYear, dateField[0].month, dateField[0].day );
          return todaysDate <= eventDate;
        }
        return true;
      } )
    } );
  }
  // add filtering logic to event listeners
  topicSelect.addEventListener( 'change', () => {
    filterLogic();
    layersUnique.add( topicSelect.value );
  } );

  topicSelect.appendChild( fragment );

  pastEventsCheckbox.addEventListener( 'change', () => {
    filterLogic();
  } );
}

// const mapData = { features: null };

// Pull map data from iip-maps API
const mapDataEndpoint = `/wp-json/iip-map/v1/maps/${mapId}`;
fetch( mapDataEndpoint )
  .then( response => response.json() )
  .then( ( mapDataData ) => {
    buildFilter( mapDataData );
    // const { features } = mapDataData;
    // store all events for use if filter is reset
    storage.all = mapDataData.features;
    // mapData.features = features;

    map.addSource( 'events', {
      type: 'geojson',
      data: mapDataData,
      cluster: true,
      clusterMaxZoom: 14,
      clusterRadius: 50
    } );

    map.addLayer( {
      id: 'clusters',
      type: 'circle',
      source: 'events',
      filter: ['has', 'point_count'],
      paint: {
        // Use step expressions (https://docs.mapbox.com/mapbox-gl-js/style-spec/#expressions-step)
        // with three steps to implement three types of circles:
        //   * Blue, 20px circles when point count is less than 100
        //   * Yellow, 30px circles when point count is between 100 and 750
        //   * Pink, 40px circles when point count is greater than or equal to 750
        'circle-color': [
          'step',
          ['get', 'point_count'],
          '#51bbd6',
          100,
          '#f1f075',
          750,
          '#f28cb1'
        ],
        'circle-radius': [
          'step',
          ['get', 'point_count'],
          20,
          100,
          30,
          750,
          40
        ]
      }
    } );

    map.addLayer( {
      id: 'cluster-count',
      type: 'symbol',
      source: 'events',
      filter: ['has', 'point_count'],
      layout: {
        'text-field': '{point_count_abbreviated}',
        'text-font': ['Arial Unicode MS Bold'],
        'text-size': 12
      }
    } );

    map.addLayer( {
      id: 'uncluster-point',
      type: 'symbol',
      source: 'events',
      filter: ['!', ['has', 'point_count']],
      layout: {
        'icon-image': 'pin',
        'icon-allow-overlap': false,
        'icon-size': 1
      }
    } );
  } )
  .catch( ( err ) => {
    console.error( err );
  } );

// convert 12 to 24 hour format if user chooses to
const convertTime12to24 = ( time12h ) => {
  const [time, modifier] = time12h.split( ' ' );
  let [hours, minutes] = time.split( ':' ); // eslint-disable-line prefer-const
  if ( hours === '12' ) {
    hours = '00';
  }
  if ( modifier === 'PM' ) {
    hours = parseInt( hours, 10 ) + 12;
  }
  return `${hours}:${minutes}`;
};

// build our sections for pop ups
const buildSection = ( field, sectionName = '' ) => {
  switch ( sectionName ) {
    case 'title':
      return card.title.toggled ? `<h3 class="info-window__title">
        ${card.title.preTitle} ${field} ${card.title.postTitle}</h3>` : '';
    case 'location':
      return card.location.toggled ? `<div class="info-window__location">
        <h4>${card.location.heading}</h4><div class="info-window__field">${field}</div></div>` : '';
    case 'date':
      if ( card.date.toggled ) {
        return `
          <h4>${card.date.heading}</h4>
          <div class="info-window__date">
              ${field[0].month}/${field[0].day}/${currentYear}
          </div>
        `.trim();
      }
      return '';
    case 'time':
      if ( card.time.toggled === true ) {
        if ( card.time.timeFormat === '24hour' ) {
          const time = convertTime12to24( `${field[0].hours}:${field[0].minutes} ${field[0].am_pm}` );
          return `<div class="info-window__time">${time}</div>`;
        }
        if ( card.time.timeFormat === '12hour' ) {
          return `<div class="info-window__time">${field[0].hours}:${field[0].minutes} ${field[0].am_pm}</div>`;
        }
      } else {
        return '';
      }
      break;
    default: return `<div class="info-window__general">${mashObject( field )}</div>`;
  }
};

// build additional data section for pop ups
// broken into it's own function due to the need to lop through mapped fields
const buildAdditional = ( field ) => {
  let fieldArr;
  let markup = '';
  const added = card.added_arr;
  if ( Array.isArray( field ) ) {
    fieldArr = field;
    // pull out field id
    added.forEach( ( o, i ) => {
      // check that field contains data before proceeding
      if ( fieldArr[i] ) {
        if ( fieldArr[i].checked ) {
          markup += `
            <div class="info-window__additional">
              <h4>${o.heading}</h4>
              <div class="info-window__field">${o.inlinePre} ${mashObject( fieldArr[i].checked )} ${o.inlinePost}</div>
            </div>
          `.trim();
        } else {
          markup += `
            <div class="info-window__additional">
              <h4>${o.heading}</h4>
              <div class="info-window__field">${o.inlinePre} ${mashObject( fieldArr[i] )} ${o.inlinePost}</div>
            </div>
          `.trim();
        }
      }
    } );
  } else {
    fieldArr = field.split( ',' );
    added.forEach( ( o, i ) => {
      markup += `
        <div class="info-window__additional">
          <h4>${o.heading}</h4>
          <div class="info-window__field">
            ${o.inlinePre} ${mashObject( fieldArr[i] )} ${o.inlinePost}
          </div>
        </div>`.trim();
    } );
  }
  return markup;
};

// on map load add clustering and popup functionality
map.on( 'load', () => {
  map.on( 'click', 'clusters', ( e ) => {
    const features = map.queryRenderedFeatures( e.point, { layers: ['clusters'] } );
    const clusterId = features[0].properties.cluster_id;
    map.getSource( 'events' ).getClusterExpansionZoom( clusterId, ( err, zoom ) => {
      if ( err ) {
        console.error( err );
        return;
      }

      map.easeTo( {
        zoom,
        center: features[0].geometry.coordinates
      } );
    } );
  } );

  // display marker on click
  map.on( 'click', 'uncluster-point', ( e ) => {
    const coordinates = e.features[0].geometry.coordinates.slice();
    const { fields } = e.features[0].properties;
    const fieldsObj = JSON.parse( fields );

    const sdid = e.features[0].properties.ext_id;

    // map selected field data to available field object
    const titleField = e.features[0].properties.title;
    const topicField = parseSection( mapping.topic_arr, fieldsObj );
    const locationField = parseSection( mapping.location_arr, fieldsObj );
    const dateField = parseSection( mapping.date_arr, fieldsObj );
    const timeField = parseSection( mapping.time_arr, fieldsObj );
    const additionalData = parseSection( mapping.other_arr, fieldsObj );
    // Ensure that if the map is zoomed out such that multiple
    // copies of the feature are visible, the popup appears
    // over the copy being pointed to.
    while ( Math.abs( e.lngLat.lng - coordinates[0] ) > 180 ) {
      coordinates[0] += e.lngLat.lng > coordinates[0] ? 360 : -360;
    }

    new mapboxgl.Popup( { offset: 25 } )
      .setLngLat( coordinates )
      .setHTML( `
                 <div class="info-window" data-sdid="${sdid}">
                   <div class="info-window__header">
                    ${buildSection( titleField, 'title' )}
                    ${buildSection( topicField, 'topic' )}
                   </div>
                   <div class="info-window__body">
                    ${buildSection( locationField, 'location' )}
                    <div class="info-window__date-time">
                      ${buildSection( dateField, 'date' )}
                      ${buildSection( timeField, 'time' )}
                    </div>
                   </div>
                   <div class="info-window__footer">
                    ${buildAdditional( additionalData )}
                   </div>
                 </div>`.trim() )
      .addTo( map );
  } );

  map.on( 'mouseenter', 'uncluster-point', () => {
    map.getCanvas().style.cursor = 'pointer';
  } );

  map.on( 'mouseleave', 'uncluster-point', () => {
    map.getCanvas().style.cursor = '';
  } );

  map.on( 'mouseenter', 'clusters', () => {
    map.getCanvas().style.cursor = 'pointer';
  } );

  map.on( 'mouseleave', 'clusters', () => {
    map.getCanvas().style.cursor = '';
  } );
} );
